### ISSUE TITLE
feat(prosemirror-keymap): Ability to specify `KeyboardEvent.code` instead of a character

### ISSUE THREAD

**Issue by Dattaya** [ProseMirror/prosemirror#1060]
### Issue details
As described here: https://discuss.prosemirror.net/t/mod-keymap-binding-on-non-english-layout-in-windows/1975, in any (most?) language when you press `ctrl+b`, even in you're not on English layout, it will make text bold. It works in all of the apps that I use (MS Office/Google Docs/etc), no one has to switch to English.
IMO, it would be nice to be able to specify `KeyboardEvent.code` or `KeyboardEvent.keyCode` instead of a character

A quote from MDN:
> The KeyboardEvent.code property represents a **physical** key on the keyboard (as opposed to the character generated by pressing the key). In other words, this property returns a value that **isn't altered by keyboard layout** or the state of the modifier keys.

https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code

A quote from the disuss thread:
> For example, in a querz layout, z and y are switched

So users of querz layout expect Ctrl+Z to work when you press Ctrl+Y (in querz layout)? I just installed it and in MS OneNote (Win10) to undo, you need to press what previously was Ctrl+Y (button in the top row)

> I think adding separate bindings for ctrl-и

Adding it for every language doesn't make much sense.

### Steps to reproduce
Add a keymap, e.g.:
```js
      keymap({
        'Mod-b': toggleMark(schema.marks.bold),
      }),
```
Switch language and press `Ctrl+b` and observe no effect.

### ProseMirror version
all

### Affected platforms

- [x] Chrome
- [x] Firefox
- [x] Internet Explorer
- [x] Other

**Comment by Dattaya** [comment]
If the first post was not clear enough, another example:
A common keyboard shortcut when you want to save something:

- Ctrl+s (English and some others) `onKeyDown` `keyCode`: 83, `code`: KeyS
- Ctrl+і (Ukrainian, Byelorussian) - 0x456; `keyCode`: 83, `code`: KeyS
- Ctrl+ы (Russian) - 0x44b; `keyCode`: 83, `code`: KeyS

And most likely there are more languages that produce neither s,і,ы when you press that button. I never have to think about it, as I use `code` and not a character.

**Comment by Dattaya** [comment]
For the time being I'll be using the workaround suggested in that discuss.prosemirror thread, it's just weird to see PM with broken i18n. Here's a workaround that Atlassian devs came up with: https://unpkg.com/browse/@atlaskit/editor-core@123.0.3/dist/es2019/utils/keymap.js

**Comment by marijnh** [comment]
> it's just weird to see PM with broken i18n.

As the thread you linked shows, I'm trying to balance several different scenarios (key maps that _should_ change key binding behavior, such as azerty or dvorak, and those that shouldn't) with a crappy browser interface (which doesn't distinguish between the two, and has plenty of other limitations).

Keymaps initially used `.code`, but that caused all kinds of problems. The kludge in that Atlaskit code is quite close to the workaround I suggested in [that thread](https://discuss.prosemirror.net/t/mod-keymap-binding-on-non-english-layout-in-windows/1975/10), but I never got further feedback on it, so it didn't get implemented.

We can continue that discussion. The fact that Atlaskit is using it might be a good indication that the blanket approach they use (anything that produces a non-ascii character should try `.code`/`.keyCode`) doesn't cause too many problems. Though I guess it would be better to first look for bindings that directly mention the character produced, and give those precedence. Does that sound reasonable to you?

**Comment by Dattaya** [comment]
> Though I guess it would be better to first look for bindings that directly mention the character produced, and give those precedence.

It's hard to compete with modern browsers that improve their `KeyboardEvent.code` handling all the time by fixing bugs and also when new input devices/layouts are introduced. I've found out that small letter 's' in Chromium source code is marked with [a comment as `// AC02`](https://github.com/chromium/chromium/blob/2ca8c5037021c9d2ecc00b787d58a31ed8fc8bcb/third_party/google_input_tools/src/chrome/os/inputview/config/us_data.js), and judging by [a github search](https://github.com/chromium/chromium/search?p=2&q=%2F%2F+AC02&unscoped_q=%2F%2F+AC02), the following letters will have identical `code`: s, ㄴ, ს, ד, ß, ș, ы, o, я, і, and many more.
Bulgarian keyboard, for example, have Cyrillic 'я' in the place of 's'. 
![image](https://user-images.githubusercontent.com/387256/82210221-4b9b7c00-9917-11ea-8483-2c0a837ffa3c.png)
So Atlaskit approach for non-ascii characters seems to be better.

**Comment by marijnh** [comment]
Does attached patch help?

**Comment by Dattaya** [comment]
Yup 👍🏻, tested in Ukrainian, Georgian, Korean, and Russian layout. 
This condition might break `name.length == 1`, but only in rare cases if the character is out of BMP, mb this never happens, not sure.

**Comment by marijnh** [comment]
Right, I considered adding a check for that, but I don't think I've seen any keyboard layout that includes non-BMP characters as direct keys yet (which doesn't mean that no such keyboard exists, of course).

**Comment by marijnh** [comment]
Released this as prosemirror-keymap 1.1.4

**Comment by Dattaya** [comment]
> Right, I considered adding a check for that, but I don't think I've seen any keyboard layout that includes non-BMP characters as direct keys yet (which doesn't mean that no such keyboard exists, of course).

Agreed, very low chance of someone using an archaic language layout, or with math symbols, or something like that, I shouldn't even have mentioned that.
Thanks!
